src/handlers/gestures/GestureMonitor.tsx

```diff
@@ -1,93 +1,399 @@
+    return true;
+  }
+  for (let i = 0; i < gesture.length; i++) {
+    if (gesture[i].handlerName !== preparedGesture.config[i].handlerName) {
+      return true;
     }
   }
 
-  private refHandler = (node: any) => {
-    const child = React.Children.only(this.props.children);
-    // TODO(TS) fix ref type
-    const { ref }: any = child;
-    if (ref !== null) {
-      if (typeof ref === 'function') {
-        ref(node);
-      } else {
-        ref.current = node;
-      }
+  return false;
+}
+
+function useAnimatedGesture(preparedGesture: GestureConfigReference) {
+  if (!Reanimated) return;
+
+  function isStateChangeEvent(
+    event:
+      | UnwrappedGestureHandlerEvent
+      | UnwrappedGestureHandlerStateChangeEvent
+  ): event is UnwrappedGestureHandlerStateChangeEvent {
+    'worklet';
+    return event.oldState != null;
+  }
+  // Hooks are called conditionally, but the condition is whether the
+  // react-native-reanimated is installed, which shouldn't change while running
+  // eslint-disable-next-line react-hooks/rules-of-hooks
+  const sharedHandlersCallbacks = Reanimated.useSharedValue<
+    HandlerCallbacks<Record<string, unknown>>[] | null
+  >(null);
+
+  const callback = (
+    event:
+      | UnwrappedGestureHandlerStateChangeEvent
+      | UnwrappedGestureHandlerEvent
+  ) => {
+    'worklet';
+
+    const currentCallback = sharedHandlersCallbacks.value;
+    if (!currentCallback) {
+      return;
     }
-  };
 
-  attachGestureHandlers(newViewTag: number) {
-    this.viewTag = newViewTag;
-    if (this.props.gesture) {
-      for (const gesture of this.props.gesture.config) {
-        if (this.props.gesture?.animatedEventHandler) {
-          RNGestureHandlerModule.attachGestureHandler(
-            gesture.handlerTag,
-            newViewTag,
-            false
-          );
+    //Using for-of loop here causes crash, because worklets do not support it,
+    //consider changing this in future when support for it is added
+    //eslint-disable-next-line @typescript-eslint/prefer-for-of
+    for (let i = 0; i < currentCallback.length; i++) {
+      const gesture = currentCallback[i];
+
+      if (event.handlerTag === gesture.handlerTag) {
+        if (isStateChangeEvent(event)) {
+          if (
+            event.oldState === State.UNDETERMINED &&
+            event.state === State.BEGAN
+          ) {
+            if (gesture.isOnBeganWorklet) {
+              gesture.onBegan?.(event);
+            } else if (gesture.onBegan) {
+              console.warn('Animated gesture callback must be a worklet');
+            }
+          } else if (
+            (event.oldState === State.BEGAN ||
+              event.oldState === State.UNDETERMINED) &&
+            event.state === State.ACTIVE
+          ) {
+            if (gesture.isOnStartWorklet) {
+              gesture.onStart?.(event);
+            } else if (gesture.onStart) {
+              console.warn('Animated gesture callback must be a worklet');
+            }
```

Comment:
Those ifs should be extracted to helper.

Reply:
Done in e3c6c64c724342f61b35c2b02a2688ac3b0af2f4